server/src/main/java/org/elasticsearch/cluster/coordination/NodeJoinExecutor.java

```diff
@@ -355,6 +360,30 @@ private static void blockForbiddenVersions(TransportVersion joiningTransportVers
         }
     }
 
+    private Set<String> calculateAllAssumedClusterFeatures(DiscoveryNodes nodes, Map<String, Set<String>> nodeFeatures) {
+        if (featureService.featuresCanBeAssumedForNodes(nodes)) {
+            Set<String> assumedFeatures = featureService.getNodeFeatures()
+                .values()
+                .stream()
+                .filter(NodeFeature::assumedAfterNextCompatibilityBoundary)
+                .map(NodeFeature::id)
+                .collect(Collectors.toSet());
+
+            // add all assumed features to the featureset of all nodes of the next major version
+            nodeFeatures = new HashMap<>(nodeFeatures);
+            for (var node : nodes.getNodes().entrySet()) {
+                if (featureService.featuresCanBeAssumedForNode(node.getValue())) {
+                    nodeFeatures.compute(node.getKey(), (k, v) -> {
+                        var newFeatures = new HashSet<>(v);
```

Comment:
The `compute` method will pass a null for `v` if the key is not present in the map. That could catch the eye of a reader of this code, since `new HashSet<>(v)` can't tolerate a null. Convincing oneself that this won't happen requires looking at a lot of code outside this method, which isn't ideal.

Anyhow, is there a reason we're not updating the sets in-place? If we could do so, the code would be simpler:

```
nodeFeatures.get(node.getKey()).addAll(assumedFeatures);
```

Reply:
How about using `computeIfPresent` instead with an additional comment mentioning that we expect an entry for every node? `calculateAllAssumedClusterFeatures` can not mutate it's input, this would break the logic following it.

Reply:
I'll use `computeIfPresent` with an assertion